### PR TITLE
8305582: Compiler crash when compiling record patterns with var

### PR DESCRIPTION
**Pull request by biboudis**
While `var` is not allowed in record pattern position the compiler was not checking it. This PR address this issue, by introducing the relevant error symbols and error types.

e.g., `if (o instanceof ColoredPoint(var(var x, var y), var c)) { }`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305582](https://bugs.openjdk.org/browse/JDK-8305582): Compiler crash when compiling record patterns with var

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13331/head:pull/13331` \
`$ git checkout pull/13331`

Update a local copy of the PR: \
`$ git checkout pull/13331` \
`$ git pull https://git.openjdk.org/jdk.git pull/13331/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13331`

View PR using the GUI difftool: \
`$ git pr show -t 13331`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13331.diff">https://git.openjdk.org/jdk/pull/13331.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13331#issuecomment-1496358176)

### CODE REVIEW

**Comment by biboudis** [comment]
Small fix, looking for a reviewer 🙏

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java [hunk 1]
```diff
@@ -4194,17 +4194,26 @@ public void visitBindingPattern(JCBindingPattern tree) {
...
-        Type type = attribType(tree.deconstructor, env);
-        if (type.isRaw() && type.tsym.getTypeParameters().nonEmpty()) {
-            Type inferred = infer.instantiatePatternType(resultInfo.pt, type.tsym);
-            if (inferred == null) {
-                log.error(tree.pos(), Errors.PatternTypeCannotInfer);
-            } else {
-                type = inferred;
+        Type site;
+
+        if (tree.deconstructor == null) {
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:4199]: my question here would be, this situation can only happen if there are `var`(s) in the record pattern?

Comment (biboudis) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:4199]: I think so, yes:

https://github.com/openjdk/jdk/blob/master/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java#L846-L869

Initially I thought I could capture this situation before line 816, but then I wouldn't be able to flow an erroneous type from parsing (`types.createErrorType(tree.record.type);`).

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:4199]: ok sounds good

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Comment by biboudis** [comment]
(do not merge, something needs fixing after the rebase)

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Review by lahodaj — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**